### PR TITLE
File/Env resolver to load app props from files/env variables

### PR DESCRIPTION
**Pull request by vijaynalawade**
**What kind of change does this PR introduce?** (check one with "x")
```
[] Bugfix
[x] Feature
[] Code style update (formatting, local variables)
[] Refactoring (no functional changes, no api changes)
[] Other... Please describe:
```

**Fixes**: #
https://github.com/TIBCOSoftware/flogo/issues/445
**What is the current behavior?**

**What is the new behavior?**

### CODE REVIEW

File: app/filevalueresolver.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+				}
+				e = json.Unmarshal(file, &props)
+				if e != nil {
+					log.Errorf("Can not read - %s due to error - %v", filePath, e)
+					panic("")
+				}
+
+				for k, v := range props {
+					preload[k] = v
+					delete(props, k)
```
Comment (debovema) [inline app/filevalueresolver.go]: Since variable props is reset at each loop iteration, why delete a key once it has been added to preload ?
Is it to avoid duplicates ?

File: app/filevalueresolver.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/TIBCOSoftware/flogo-lib/logger"
+)
+
+var preload = make(map[string]interface{})
+
+var log = logger.GetLogger("app-props-resolver-file")
+
+// Comma separated list of json files overriding default application property values
+// e.g. FLOGO_APP_PROPS_FILE_CONFIG=app1.json,common.json
+const EnvAppPropertyFileConfigKey = "FLOGO_APP_PROPS_FILE_CONFIG"
```
Comment (debovema) [inline app/filevalueresolver.go]: Since we are somehow redesigning the configuration management in Flogo, I wonder if existing prefixes are good.
* **FLOGO_** : OK
* **APP_**: is it really useful ? all other environment variables such as "FLOGO_DATA_SECRET_KEY", "FLOGO_PUBLISH_AUDIT_EVENTS", "FLOGO_LOG_LEVEL"... target a Flogo Application. I think we should drop this from the prefix for configuration mangement for simplicity's sake :)
* However since we are in the configuration management scope, we could add a **CONFIG_** prefix instead
* Lastly, the **PROPS_FILE** is misleading since the expected format is not a standard key=value properties file but a JSON file. It would be better to use something more explicit like **PROFILES**. The documentation would explain what a *profile* is in the Flogo context. It's consistent with other tools like BW for instance and the trailing **S** emphasizes the fact that the variable can contain multiple values (comma-separated list of *profiles*).

To sum up, I would suggest the following environment variable name: **FLOGO_CONFIG_PROFILES**

File: app/filevalueresolver.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+func init() {
+
+	filePaths := getExternalFiles()
+	if filePaths != "" {
+		// Register value resolver
+		RegisterPropertyValueResolver("file", &FileValueResolver{})
+
+		if config.GetAppPropertiesValueResolver() == "" {
+			//Make file resolver default since FLOGO_APP_PROPS_FILE_CONFIG is set
```
Comment (debovema) [inline app/filevalueresolver.go]: I don't understand why the fact that the environment variable is set implies that the resolver becomes the default property resolver. Will it be the same for other resolvers ? How the priorities will be managed ?

Comment (vijaynalawade) [reply app/filevalueresolver.go]: @debovema : This is for the developer's convenience. Note that I am setting it implicitly only when _FLOGO_APP_PROPS_VALUE_RESOLVER_ is not explicitly set. I dont think we should trouble developers for simple configuration like value resolver when intention is to use only one resolver like file or consul or env.
If developers intend to use multiple of those (chaining), then they should explicitly set _FLOGO_APP_PROPS_VALUE_RESOLVER_.

File: app/filevalueresolver.go (hunk 2 above)
Comment (vijaynalawade) [reply app/filevalueresolver.go]: @debovema : I prefer more explicit naming to give developers context of the configuration. In my opinion, FLOGO_CONFIG_PROFILES does not give clear indication of what kind configuration. Is it engine, app props or something else?
I think, we should follow consistent naming like FLOGO_APP_PROPS_FILE_OVERRIDE (for file resolver), FLOGO_APP_PROPS_CONSUL_OVERRIDE (for Consul resolver), FLOGO_APP_PROPS_ENV_OVERRIDE (for env resolver). This kind of naming give me clear indication of what it really does.

File: app/filevalueresolver.go (hunk 1 above)
Comment (vijaynalawade) [reply app/filevalueresolver.go]: @debovema I forgot to remove that. Thanks for pointing it out.

File: app/filevalueresolver.go (hunk 2 above)
Comment (emilienthomas) [reply app/filevalueresolver.go]: In addition to @debovema comment, I feel that the FLOGO_CONFIG prefix is locked because  
FLOGO_CONFIG_PATH is used to point the flogo.json that defines the application.
What about renaming FLOGO_CONFIG_PATH to FLOGO_APP_PATH, then use FLOGO_CONFIG prefix for configuration management ?

Comment (mellistibco) [reply app/filevalueresolver.go]: I don't think `FLOGO_APP_PROPS_FILE_OVERRIDE` is great... However, at the same time, overloading the term config may become a bit of a challenge, as the app.json, in itself, is really just a config file for the engine... Inside this config, we have the notion of properties... So in reality, you're using (potentially) an external configuration mgmt tool to override properties within your flogo app config... We surface the notion of application just about everywhere, however, so I suspect differentiating "engine" config vs "app" config may make sense.

@emilienthomas suggestion in renaming `FLOGO_CONFIG_PATH` to `FLOGO_APP_PATH` may make sense... However, I'd still suggest that keeping everything properly scoped is important... So, regardless of that change, you'd still probably want to have something like `FLOGO_APP_CONFIG_PROPERTY_RESOLVER` or something, which begins to get a bit long imo...

File: app/filevalueresolver.go (hunk 3 above)
Comment (mellistibco) [reply app/filevalueresolver.go]: 🤔 I kind of tend to lean toward not setting a default... I think it may make more sense for them to explicitly set the resolver to `file` not just assume that's what they want if file name(s) have been specified in the other env var... Could make troubleshooting a bit more of a challenge as we're inferring behavior which may not be what is expected... I'd rather stay away from this. If the file resolver isn't set, regardless of them specifying files in another env var, then the default resolver is used (or none at all)...

**Comment by mellistibco** [comment]
Also not a huge fan of the file name... `filevalueresolver`... Perhaps this should be inside another package rather than just under `app`? like `app/resolver` and then the name could just be `file` or something along those lines... @fm-tibco ?

**Comment by vijaynalawade** [comment]
@mellistibco : We can move this under **app/propertyresolver/file.go**.

File: app/filevalueresolver.go (hunk 3 above)
Comment (vijaynalawade) [reply app/filevalueresolver.go]: Alright. Will remove it for now.

File: app/filevalueresolver.go (hunk 2 above)
Comment (vijaynalawade) [reply app/filevalueresolver.go]: How about following?
```
FLOGO_APP_PROPS_CONFIG_FILE
FLOGO_APP_PROPS_CONFIG_CONSUL
FLOGO_APP_PROPS_CONFIG_ENV
```

**Comment by vijaynalawade** [comment]
@mellistibco : Check updated PR. Moved file resolver under `app/propertyresolver`

File: app/filevalueresolver.go (hunk 2 above)
Comment (vijaynalawade) [reply app/filevalueresolver.go]: @debovema: As part of PR I reused your implementation for env resolver. This will help us find better and consistent naming. :-)

**Comment by fm-tibco** [comment]
Just so summarize, we don't want to use 

FLOGO_APP_PROPS_VALUE_RESOLVER=consul,env,file

To specify what resolver to use and in what order?  Basically a resolver will be enabled if it's corresponding "CONFIG" env variable is used.  

As for $env do we want to keep that?  Given that we have an ENV property resolver?  That way the developer doesn't have to think in advance how his properties will be resolved.  That can be done at the end without having to modify their application.  We can make the default resolver the ENV one.

**Comment by fm-tibco** [comment]
Maybe instead of override, it can be alias.  So a mapping of the internal name to the external if needed (like consul or env). So the external config mapping becomes an ops exercise.  The question is can there just be one alias env var, or do we need on for each resolver type.  I worry that this might get too complicated.

We need to make sure the simplest use case is simple.

**Comment by vijaynalawade** [comment]
@fm-tibco : I think alias mechanism won't be that common. So, let's not worry about that. 
I still expect single alias for all resolver types but current mechanism can be used to create different aliases for different resolver types and environments.
e.g. In dev env, ```FLOGO_APP_PROPS_OVERRIDE={"workdir": "$HOME"} FLOGO_APP_PROPS_CONFIG_ENV=true```
In test env,  ```FLOGO_APP_PROPS_OVERRIDE={"workdir": "$/common/workdir"} FLOGO_APP_PROPS_CONFIG_CONSUL={"Address": "http://localhost:8500"}```

**Comment by fm-tibco** [comment]
How about something super simple:

```
FLOGO_PROPS_RESOLVER = consul,env,json

FLOGO_PROPS_JSON = { "value": "one" }
or
FLOGO_PROPS_JSON = /tmp/props.json

FLOGO_PROPS_ENV  = { "workingDir": "WORKING_DIR" } //optional prop aliases

FLOGO_PROPS_CONSUL = {"service_address": "http://localhost:8500", "aliases":{"val":"/a/b/c"}}
or
FLOGO_PROPS_CONSUL = /tmp/consul.config
```
Basically if aliases are needed (like in consul or env), they can be part of its corresponding config

We could add a CONFIG suffix if we like

**Comment by vijaynalawade** [comment]
@fm-tibco : I like the alias idea.
I still feel APP is needed in the name as there are some engine properties too. More explicit better from configuration perspective.

FLOGO_APP_PROPS_RESOLVER = consul,env,json

FLOGO_APP_PROPS_JSON = { "value": "one" }
or
FLOGO_APP_PROPS_JSON = /tmp/props.json

FLOGO_APP_PROPS_ENV  = { "workingDir": "WORKING_DIR" } //optional prop aliases

FLOGO_APP_PROPS_CONSUL = {"service_address": "http://localhost:8500", "aliases":{"val":"/a/b/c"}}
or
FLOGO_APP_PROPS_CONSUL = /tmp/consul.config

**Comment by debovema** [comment]
Here are my thoughts on last proposition of @fm-tibco :

# FLOGO_PROPS_RESOLVER
```
FLOGO_PROPS_RESOLVER = consul,env,json
```

OK but do we really need to do something like :
```
FLOGO_PROPS_RESOLVER = etcd,redis,consul,env,json
```

I mean there should be only one "external" resolver and "env" and "json" could be always there as default.

So either change to:
```
FLOGO_PROPS_RESOLVER = external,env,json
```
if we want to define the priority order between the resolvers (external and builtins),

or no ```FLOGO_PROPS_RESOLVER``` at all and the priority order is hardcoded in the flogo-lib code.

# FLOGO_APP_PROPS_CONSUL
```
FLOGO_APP_PROPS_CONSUL
```
Is this variable included in the flogo-lib repository ?
IMO it should not be. What if we want to implement config management backed by etcd, by redis, etc... ? 

The configuration can be provided by environment variable(s) defined by the external resolver themselves.

**Comment by fm-tibco** [comment]
I am fine with just having one external provider at a time, it simplifies things.

FLOGO_PROPS_RESOLVER = external,env,json

I am also fine with not really requiring this resolver property and not having to specify it, maybe we can have people optionally specify it if they really want to change the order.

As for FLOGO_APP_PROPS_CONSUL, my intent wasn't for the lib to specify that in any way.  I was just using it as an example since it has been mentioned so many times here.  I expect the "external" resolver to designate this variable.

**Comment by mellistibco** [comment]
I would agree, I don't think having multiple external providers is something that many people would really do, but at the same time, do we really care how many config providers they want to stack together?

**Comment by fm-tibco** [comment]
I'm fine with that as well

**Comment by vijaynalawade** [comment]
@mellistibco @debovema @fm-tibco : To sum up, I assume we agreed on below:

FLOGO_APP_PROPS_RESOLVER = consul,env,json (As @mellistibco said, everything is considered external and explicit.)

**JSON Resolver (part of core)**
FLOGO_APP_PROPS_JSON = { "value": "one" }
or
FLOGO_APP_PROPS_JSON = /tmp/props.json

**Env Resolver(part of core)**
FLOGO_APP_PROPS_ENV = {"overrides": {"<propName>": "<Overridden env Name>"}}. (Optional but very useful when app props names don't correspond to env var names. **FLOGO_APP_PROPS_RESOLVER=env** must be set.)

**Comment by debovema** [comment]
LGTM
Can we assume there will be a default value of "env,json" for FLOGO_APP_PROPS_RESOLVER ?

And other resolvers can implement their own configuration and behaviour of course.

File: app/propertyresolver/file.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package propertyresolver
```
Comment (fm-tibco) [inline app/propertyresolver/file.go]: should this file be renamed to json.go?

Comment (vijaynalawade) [reply app/propertyresolver/file.go]: @fm-tibco : Fixed in latest PR.

**Comment by vijaynalawade** [comment]
@fm-tibco : Current PR does not cover chaining capability. I will create separate PR for that. No default set for FLOGO_APP_PROPS_RESOLVER. So, developers need to set it based on their use cases.

File: app/filevalueresolver.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		// preload props from files
+		files := strings.Split(filePaths, ",")
+		if len(files) > 0 {
+			for _, filePath := range files {
+				props := make(map[string]interface{})
+
+				file, e := ioutil.ReadFile(filePath)
+				if e != nil {
+					log.Errorf("Can not read - %s due to error - %v", filePath, e)
+					panic("")
```
[the remark this answers is not in the document]
Comment (vijaynalawade) [reply app/filevalueresolver.go]: @mellistibco : Most log based tooling look for **ERROR** phrase for alerting. Hence, I opt to log error and just panic. In this case, panic is needed because user intend to use this mode through env var but did not configure it properly.
